File: apps/web/src/server/services/timer.ts
import { addMilliseconds, isBefore, differenceInSeconds } from 'date-fns'
import { type Settings, type Timer, TimerStatus } from '@/types/types'

export function calculateCurrentTimer({ settings, timer }: {
  settings: Settings
  timer: Timer
}) {
  const now = new Date()
  const state: Timer = {
    startTime: timer.startTime,
    status: timer.status,
    successionCount: timer.successionCount,
    totalWorkTime: timer.totalWorkTime,
    totalBreakTime: timer.totalBreakTime,
    elapsedPrePause: 0,
  }
  while (true) {
    if (state.status === TimerStatus.Working) {
      const endTime = addMilliseconds(state.startTime, settings.workLength * 1000)
      if (isBefore(endTime, now)) {
        if (state.successionCount <= 1) {
          state.status = TimerStatus.LongBreak
          state.successionCount = settings.breakSuccessions
        }
        else {
          state.status = TimerStatus.ShortBreak
          state.successionCount -= 1
        }
        state.totalWorkTime += settings.workLength
        state.startTime = endTime
        continue
      }
      state.totalWorkTime += settings.workLength - differenceInSeconds(endTime, now)
    }
    else if (state.status === TimerStatus.ShortBreak) {
      const endTime = addMilliseconds(state.startTime, settings.shortBreakLength * 1000)
      if (isBefore(endTime, now)) {
        state.status = TimerStatus.Working
        state.startTime = endTime
        state.totalBreakTime += settings.shortBreakLength
        continue
      }
      state.totalBreakTime += settings.shortBreakLength - differenceInSeconds(endTime, now)
    }
    else if (state.status === TimerStatus.LongBreak) {
      const endTime = addMilliseconds(state.startTime, settings.longBreakLength * 1000)
      if (isBefore(endTime, now)) {
        state.status = TimerStatus.Working
        state.startTime = endTime
        state.totalBreakTime += settings.longBreakLength
        continue
      }
      state.totalBreakTime += settings.longBreakLength - differenceInSeconds(endTime, now)
    }
    return state
  }
}

File: apps/mobile/lib/providers/timer.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:pomo_mobile/models/timer/timer.dart';
import 'package:pomo_mobile/providers/auth.dart' show authRequestProvider;
import 'package:pomo_mobile/providers/settings.dart';
import 'package:pomo_mobile/services/timer_service.dart';

final timerProvider = StateProvider<Timer?>((_) => null);

final countdownProvider = StateProvider<int>((_) => 0);

final calculateTimeProvider = Provider((ref) {
  return ({Timer? current}) {
    final timer = current ?? ref.read(timerProvider);
    final settings = ref.read(settingsProvider);
    if (timer == null || settings == null) return;
    if (timer.status == TimerStatus.paused) {
      return;
    }
    var length = switch (timer.status) {
      (TimerStatus.working) => settings.workLength,
      (TimerStatus.longBreak) => settings.longBreakLength,
      (TimerStatus.shortBreak) => settings.shortBreakLength,
      _ => 0,
    };
    final statusEndTime = timer.startTime.add(Duration(seconds: length));
    final difference = statusEndTime.difference(DateTime.now());
    ref.read(countdownProvider.notifier).state = difference.inSeconds;
  };
});

final tickProvider = Provider((ref) {
  return () {
    final timer = ref.read(timerProvider);
    final settings = ref.read(settingsProvider);
    final countdown = ref.read(countdownProvider);
    if (timer == null || settings == null) return;
    final calculateTime = ref.read(calculateTimeProvider);

    TimerStatus status = timer.status;
    int successionCount = timer.successionCount;
    if (status == TimerStatus.working) {
      if (countdown <= 0) {
        if (successionCount <= 1) {
          status = TimerStatus.longBreak;
          successionCount = settings.breakSuccessions;
          // TODO: notify
        } else {
          status = TimerStatus.shortBreak;
          successionCount = successionCount - 1;
        }
        final current = timer.copyWith(
          startTime: timer.startTime.add(
            Duration(seconds: settings.workLength),
          ),
          status: status,
          successionCount: successionCount,
        );
        ref.read(timerProvider.notifier).state = current;
        calculateTime(current: current);
        return;
      }
      calculateTime();
      return;
    }
    if (status == TimerStatus.longBreak || status == TimerStatus.shortBreak) {
      if (countdown <= 0) {
        // TODO: notify
        final current = timer.copyWith(
          startTime: timer.startTime.add(
            Duration(
              seconds:
                  status == TimerStatus.longBreak
                      ? settings.longBreakLength
                      : settings.shortBreakLength,
            ),
          ),
          status: TimerStatus.working,
        );
        ref.read(timerProvider.notifier).state = current;
        calculateTime(current: current);
        return;
      }
      calculateTime();
    }
  };
});

final timerServiceProvider = Provider((ref) {
  final request = ref.watch(authRequestProvider);
  if (request == null) return null;
  return TimerService(request: request);
});

final getTimerProvider = Provider((ref) {
  return () async {
    final timerService = ref.read(timerServiceProvider);
    if (timerService == null) {
      return;
    }
    final timer = await timerService.fetchTimer();
    if (timer == null) {
      return;
    }
    ref.read(timerProvider.notifier).state = timer;
    ref.read(calculateTimeProvider)();
  };
});
